File: Week01/d4_combination.c
/**
 * Author: Penut Chen
 * Date: 2019-6-30
 * Description: 
 *   Practice of combination function.
 */

#include <stdio.h>
#include <stdlib.h>

int* generate_array(int);
void combination(int*, int, int, int, int*, int*);

int main() {
    int size = 5, length = 3, count = 0;
    int *arr = generate_array(size);
    int *copy = (int*)malloc(sizeof(int) * size);

    combination(arr, size, 0, length, copy, &count);
    printf("Total %d", count);
    free(arr);

    return 0;
}

int* generate_array(int size) {
    int *arr = (int*)malloc(sizeof(int) * size);
    for (int i = 0; i < size; i++)
        arr[i] = i + 1;
    return arr;
}

void combination(int *arr, int size, int idx, int length, int *copy, int *count) {
    if (idx == length) {
        for (int i = 0; i < length; i++)
            printf("%d ", copy[i]);
        printf("\n");
        *count += 1;
        return ;
    }

    for (int i = 0; i < size; i++) {
        copy[idx] = arr[i];
        combination(arr, size, idx + 1, length, copy, count);
    }
}

File: Week01/d4_combination.py
"""
Author: Penut Chen
Date: 2019-6-30
Description:
  Practice of combination function.
"""

def main():
    Combination(['True', 'False'], 5)

class Combination:
    def __init__(self, arr, length):
        self.arr = [str(a) for a in arr]
        self.length = length
        self.copy = list(range(length))
        self.size = len(arr)
        self.count = 0
        self.combination(0)
        print('Total %d' % self.count)
    
    def combination(self, idx):
        if idx == self.length:
            print(' '.join(self.copy))
            self.count += 1
            return
        
        for a in self.arr:
            self.copy[idx] = a
            self.combination(idx + 1)

if __name__ == "__main__":
    main()
